[coq] Allow dependencies and composition with local Coq libraries.

ejgallego opened the pull request:
We support simple composition of Coq libraries using the
 `(theories ...)` field. The main changes are:

- Coq libraries now have a `Coq_lib.t` type

- Stanza scanning does now initialize a `Coq_lib.DB`, per
  scope. Supporting inter-scope public libraries is left for a future
  PR.

- We properly compose when the stdlib is in scope, this allows to have
  a compositional build with Coq itself; note that as of today, in a
  composed build the stdlib prefix `Coq` is introduced qualified; this
  reflects upstream roadmap of deprecating implicit libs. See comment
  in `Scope.public_libs`

We have added tests for:

 - simple composition
 - composition with plugin
 - cyclic composition
 - boot composition [disabled until Coq 8.12 is released]
 - inter-scope composition [fails for now]

We leave for future work:

- Allow to depend on theories in a different scope
- Allow to depend on theories installed globally, likely this will
  require the 2.0 version of the Coq language
- Validation of Coq library names should be improved [clarify upstream
  rules]

Includes changes by Rudi Grinberg [many suggestions] and Théo
Zimmerman [documentation review]

ejgallego committed:
[coq] Allow dependencies and composition with local Coq libraries.

We support simple composition of Coq libraries using the
 `(theories ...)` field. The main changes are:

- Coq libraries now have a `Coq_lib.t` type

- Stanza scanning does now initialize a `Coq_lib.DB`, per
  scope. Supporting inter-scope public libraries is left for a future
  PR.

- We properly compose when the stdlib is in scope, this allows to have
  a compositional build with Coq itself; note that as of today, in a
  composed build the stdlib prefix `Coq` is introduced qualified; this
  reflects upstream roadmap of deprecating implicit libs. See comment
  in `Scope.public_libs`

We have added tests for:

 - simple composition
 - composition with plugin
 - cyclic composition
 - boot composition [disabled until Coq 8.12 is released]
 - inter-scope composition [fails for now]

We leave for future work:

- Allow to depend on theories in a different scope
- Allow to depend on theories installed globally, likely this will
  require the 2.0 version of the Coq language
- Validation of Coq library names should be improved [clarify upstream
  rules]

Includes changes by Rudi Grinberg [many suggestions] and Théo
Zimmerman [documentation review]

Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

rgrinberg commented on src/coq_lib.ml:
```diff
@@ -0,0 +1,48 @@
+
+type t =
+  { name : Lib_name.t
+  ; src_dir : Path.t
```
Perhaps a bit descriptive would be `src_root`. AFAIK, coq libs can span more than one source dir. Yes, we have `src_dir` on the OCaml side, but that is mostly an artifact of the days where only one src dir was allowed.

ejgallego replied:
Why not? Changed to root.

rgrinberg commented on src/coq_lib.mli:
```diff
@@ -0,0 +1,27 @@
+
+  val resolve : t -> Loc.t * Lib_name.t -> lib Or_exn.t
+
+end
```
You can add `with lib := t` and change `type lib = t` to `type lib`. This will relieve you from adding the not very useful type alias in the implementation.

ejgallego replied:
Done thanks.

rgrinberg commented on src/coq_rules.ml:
```diff
@@ -66,7 +71,9 @@ let setup_rule ~expander ~dir ~cc ~source_rule ~wrapper_name
-  let iflags = Arg_spec.As ["-R"; "."; wrapper_name] in
+  let src_dir = Path.drop_build_context_exn dir in
+  let th_flags = List.concat_map theories_libs ~f:setup_theory_flag in
+  let iflags = Arg_spec.As (["-R"; Path.to_string src_dir; wrapper_name] @ th_flags) in
```
I believe you can use the `Path` constructor in `Arg_spec` in conjunction with `S`. to avoid the path stringification.

ejgallego replied:
Done, thanks; improved a bit the general flag setup; still can be better tho.

ejgallego replied:
Umm, I had to revert the use of `Arg_spec.Path` as it was producing wrong paths to `-Q`, instead of `-Q ssreflect` I am getting `-Q ../../ssreflect`. I will investigate more.

ejgallego replied:
Oh Indeed, it was my fault, `Path` handles `Build.run` `~dir` properly so I had to drop the drop_build_context hack. At some point I was aware of this but I forgot. Path code looks clean now IMO.

rgrinberg commented on src/coq_rules.ml:
```diff
@@ -144,40 +153,56 @@ let setup_ml_deps ~scope ~loc libs =
+  List.concat_map ~f:(fun lib_dep ->
+    let loc = Dune_file.Lib_dep.get_loc lib_dep in
+    Dune_file.Lib_dep.to_lib_names lib_dep
+    |> List.map ~f:(fun name -> Result.ok_exn (Coq_lib.DB.resolve lib_db (loc, name))))
```
I think it would be nicer if we had a `find_many` in `Coq_lib.DB` that took a list of names.

ejgallego replied:
Good idea, done!

rgrinberg commented on src/dune_file.ml:
```diff
@@ -1842,18 +1849,22 @@ module Coq = struct
          let (loc, res) = name in
-         (loc, Lib_name.Local.validate (loc, res) ~wrapped:None)
+         match res with
+         | Ok t | Warn t -> (loc, t)
```
You should fail on `Warn` here as well. Dune does not fail because of backwards compatibility issues. There's no reason for the coq plugin to inherit that.

By the way, this is why I think it's a good reason to avoid reusing `Lib_name` at all on the coq side. You should really rely on your own validation.

ejgallego replied:
Indeed this is pending; the time we stop piggybacking on `Lib_name` has come; at least in this version of the language, we need to be able to name Coq libs such that they share some common root such as `(name foo.bar)` `(name foo.baz)`.

ejgallego commented:
Ok, so I am reasonably happy with this; testing seems to be doing fine and the code is not too horrible.

A big todo is indeed to define the notion of coq libraries names, this is a slightly complex issue so I'll think about it a couple of days [as the library names are really wrappers and there can be overlapping].

IMHO I think it does make sense to add both closure and the possibility to depend on globally-installed libs on this PR; no point on having the users go thru incomplete steps.

rgrinberg commented on src/coq_rules.ml:
```diff
@@ -112,14 +121,15 @@ let create_ccoq sctx ~dir =
+let libs_of_coq_deps ~lib_db =
+  List.concat_map ~f:(fun lib_dep ->
+    let loc = Dune_file.Lib_dep.get_loc lib_dep in
+    Dune_file.Lib_dep.to_lib_names lib_dep
```
I should have mentioned this earlier, but this isn't really handling `select` dependencies correctly. IMO, if such a feature isn't strictly necessary in coq then it shouldn't be supported. The way it's done for OCaml is annoyingly complicated, so if coq can do without it, that would be ideal. Otherwise, it should be properly supported as the current behavior is just confusing.

ejgallego replied:
Good point, I am indeed not sure we want to support `select` dependencies, I will switch the implementation to direct-style for now.

ejgallego replied:
Submitting a different PR with improvements here.

rgrinberg commented on src/super_context.ml:
```diff
@@ -287,23 +287,16 @@ let create
+          | Library lib -> Left (ctx_dir, lib)
+          | Coq.T coq_lib -> Right (ctx_dir, coq_lib)
+          | _ -> Skip))
+    |> List.split |> (fun (ll,lr) -> List.concat ll, List.concat lr)
```
If this is about improving performance, then I think most of the gains will be lost by the subsequent split and concatenations. I suggest to just write this with folds if you think it's important to make this fast. It could gain some clarity as well I think.

ejgallego replied:
I dunno what to do here, the original code was using `concat_map` which is also quite unoptimal; more than performance I was worried about keeping a "single pass" over the stanzas, somehow it should be a bit more generic than we have now indeed; I'll give it a go.

rgrinberg replied:
I just replaced the code with two folds. See the commit.

Zimmi48 commented on doc/coq.rst:
```diff
@@ -32,35 +33,74 @@ The basic form for defining Coq libraries is very similar to the OCaml form:
-  make-based Coq package eco-system. For compatibility, we also installs the `.cmxs`
-  files appearing in `<ocaml-librarie>` under the `user-contrib` prefix.
-- ``<coq_flags>`` will be passed to ``coqc``,
+The stanza will build all ``.v`` files on the given directory. The semantics of fields is:
```
```suggestion
The stanza will build all ``.v`` files in the given directory. The semantics of fields is:
```

Zimmi48 commented on doc/coq.rst:
```diff
@@ -32,35 +33,74 @@ The basic form for defining Coq libraries is very similar to the OCaml form:
+  ``foo.Bar.module1``, ``foo.Bar.module2``, etc... Note that modules
+  in the same theory don't see the ``foo.Bar`` prefix, in the same
+  way that OCaml ``wrapped`` libraries do. For compatibility reasons,
+  the 1.0 version of the Coq language installs a theory named
```
Doesn't exist yet.

ejgallego replied:
Thansk for the review @Zimmi48 ; I've added a note about 1.0.

Zimmi48 commented on doc/coq.rst:
```diff
@@ -32,35 +33,74 @@ The basic form for defining Coq libraries is very similar to the OCaml form:
+- ``public_name`` is of the form ``pkg_name.identifier``; if present,
+  Dune will generate install rules for the ``.vo`` files on the
+  theory. ``pkg_name`` must be a valid package name; ``identifier``
+  is not used in the 1.0 version of the language as in the legacy
```
Ditto

Zimmi48 commented on doc/coq.rst:
```diff
@@ -32,35 +33,74 @@ The basic form for defining Coq libraries is very similar to the OCaml form:
+  is not used in the 1.0 version of the language as in the legacy
+  install setup, all packages share a common root namespace. Indeed,
+  files will be installed in ``lib/coq/user-contrib/<module_prefix>``,
+  as customary in the make-based Coq package eco-system. For
```
```suggestion
  as customary in the make-based Coq package ecosystem. For
```

Zimmi48 commented on doc/coq.rst:
```diff
@@ -32,35 +33,74 @@ The basic form for defining Coq libraries is very similar to the OCaml form:
+  theory to depend on a ML plugin,
+- your Coq theory can depend on other theories by specifying them in
+  the ``<coq_libraries>`` field; if the theory is in-tree, you can use
+  either the public or a ``<module_prefix>``; Dune will then pass to
```
```suggestion
  either the public name or a ``<module_prefix>``; Dune will then pass to
```

?

Zimmi48 commented on doc/coq.rst:
```diff
@@ -32,35 +33,74 @@ The basic form for defining Coq libraries is very similar to the OCaml form:
+  the ``<coq_libraries>`` field; if the theory is in-tree, you can use
+  either the public or a ``<module_prefix>``; Dune will then pass to
+  Coq the corresponding flags to make everything compile correctly
+  [``-Q``]. For the 1.0 version of the Coq language, if the library is
```
Ditto

Zimmi48 reviewed (commented):
The feature makes sense like this. The doc is mostly fine.

Zimmi48 commented:
> I've added a note about 1.0.

Did you forget to push it?

ejgallego commented:
> Did you forget to push it?

I was rewriting it, sorry; note added now.

Zimmi48 commented:
Looks good, thanks!

rgrinberg commented on src/coq_lib.ml:
```diff
@@ -0,0 +1,55 @@
+  { name : Coq_lib_name.t
+  ; wrapper : string
+  ; src_root : Path.t
+  ; obj_root : Path.t
```
Sorry to be picky, but perhaps this should be in `obj_dir`? Or is possible to have more than one directory for artifacts in a coq lib? In OCaml, we can have private modules which requires us to make this separation. Doesn't seem like it's the case for coq though.

ejgallego replied:
What do you mean @rgrinberg ? I named it root after your advice indeed as qualified libs do provide qualified object dirs in the current version of Coq.

p.s: don't bother pushing to the PR, my current version has a very large delta, in particular it doesn't use `Lib_name.t` anymore.

rgrinberg replied:
I see. Well ignore my comment then. I thought the situation was analogous to OCaml where there's only one obj dir (ignoring private modules), but many source directories.

ejgallego commented:
> p.s: don't bother pushing to the PR, my current version has a very large delta, in particular it doesn't use `Lib_name.t` anymore.

I will signal when this is ready for review again; after some thinking I am convinced that it makes no sense to support `public_name` in the 1.0 version of the language, as the subcomponent has really no meaning given that current install namespace is global.

Thus, I will deprecate `public_name` and require instead a `package` field. We will hopefully able to improve this in the 2.0 version of the language.

cpitclaudel commented:
Oh, this would be great to have — I've been struggling with this.  Anything large needed before this can be merged?

rgrinberg commented on src/dune/coq_rules.ml:
```diff
@@ -192,31 +201,60 @@ let setup_ml_deps ~lib_db libs =
+  Dir_contents.coq_modules_of_library dir_contents ~name
+
+let coq_theories_of_lib_deps ~coq_lib_db libs =
+  Result.List.map ~f:(Coq_lib.DB.resolve coq_lib_db) libs |> Result.ok_exn
```
As discussed offline, this is the only bit that needs to be improved before merging.

ejgallego replied:
Finally I've opted not to delay the resolution here (yet); this is needed for more than flags, in particular for the rules for mlpack and sources, so we gain little from delaying here (unfortunately)

rgrinberg commented on src/dune/lib.ml:
```diff
@@ -1655,6 +1655,7 @@ module DB = struct
       | Library of Path.Build.t * Dune_file.Library.t
       | External_variant of Dune_file.External_variant.t
       | Deprecated_library_name of Dune_file.Deprecated_library_name.t
+      | CoqTheory of Path.Build.t * Dune_file.Coq.t
```
Should be `CoqTheory`.

ejgallego commented:
Hi folks! It took a while but IMO this is finally ready for merging.

I think code in some parts can be cleaned up a bit, I plan to keep working on this to provide inter-scope composition which is the main next feature we want. I've opted not to implement that yet as in-scope composition is enough to (finally!!) bootstrap Coq.

Testing so far seems pretty positive.

ejgallego commented:
More generally, I think we can refactor the scope code to handle libraries on different languages in less of an ad-hoc way, I will try something in the PR for the inter-scope composition.

ejgallego commented:
I see I'm late for 2.4 by very little O_O @rgrinberg should I move the change entry to 2.4.1 or directly to 2.5?

Zimmi48 commented:
Even if `(boot)` is mostly for internal use, it would seem better if it were documented, no?

ejgallego commented:
> Even if `(boot)` is mostly for internal use, it would seem better if it were documented, no?

I'm reluctant to have it documented yet, it is too unstable. Note that it was not introduced in this PR.

ejgallego commented:
Rebased, @rgrinberg I'll wait to merge until your change request is lifted.

ejgallego committed:
[coq] Refactor Coq_lib errors messages

We place them into a module; error handling in this part could
indeed use more work but this is a first step.

Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[coq] Don't allow public theories to depend on private ones.

For now we follow very simple rules, public theories cannot depend on
any way on private ones.

Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego commented:
Added a couple of cleanups + check on private vs public libs.

rgrinberg commented on src/dune/coq_lib.ml:
```diff
@@ -0,0 +1,152 @@
+  let find_many t ~loc = Result.List.map ~f:(fun name -> resolve t (loc, name))
+
+  (* Where should we move this? *)
+  module Result_monad : Monad_intf.S with type 'a t = ('a, exn) Result.t =
```
This is just `Or_exn`  I believe.

ejgallego replied:
Fixed thanks.

rgrinberg commented on src/dune/coq_lib.ml:
```diff
@@ -0,0 +1,152 @@
+    let ( >>= ) = Result.O.( >>= )
+  end
+
+  module Coq_lib_closure = Top_closure.Make (String.Set) (Result_monad)
```
Why don't you use `Result`?

ejgallego replied:
You mean in place of `Result_monad`? I didn't find a place with the right signature; I could adapt the code tho to add this to stdune but I thought we'd discuss it first.

rgrinberg replied:
Yeah, agreed. I ended up replacing it with `Or_exn` by adding the appropriate functions to `Or_exn`.

rgrinberg commented on src/dune/coq_lib_name.ml:
```diff
@@ -12,13 +12,16 @@ let to_string x = String.concat ~sep:"." x
  fun lib -> Dune_lang.Encoder.string (to_string lib)
 
-(* let pp x = Pp.text (to_string x) *)
+let pp fmt x = Fmt.text fmt (to_string x)
```
Why can't you use `pp`?

ejgallego replied:
I don't remember why I did this change, reverted.

rgrinberg commented on src/dune/coq_rules.ml:
```diff
@@ -74,42 +93,43 @@ let parse_coqdep ~boot_type ~coq_module (lines : string list) =
-        (Some "Init")
+      String.equal (Coq_lib.wrapper lib) wrapper_name
+      && Option.equal String.equal
+           (List.nth_opt (Coq_module.prefix coq_module) 0)
```
`List.hd`?

ejgallego replied:
Indeed, fixed thanks.

rgrinberg commented:
I had a first pass. Everything looks mostly fine. I think this should be ready for 2.5

ejgallego commented:
> I had a first pass. Everything looks mostly fine. I think this is good to go for 2.5

Thanks @rgrinberg , comments addressed, except for the one on `Result_monad` which I'm not sure what should be done.

I'm gonna submit the second part of this, supporting inter-scope composition, that should hopefully make the construction of the `Coq_lib.DB` less intrusive and a bit more modular.

rgrinberg committed:
Clean up monad handling

Make it more like the way we handle applicative

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented:
Looks good so far. I'd like to see a lot of the DB handling stuff to be consistent between coq & ocaml, but I suppose there's no rush.

ejgallego commented:
Thanks @rgrinberg !

> I'd like to see a lot of the DB handling stuff to be consistent between coq & ocaml, but I suppose there's no rush.

For now I'm going to separate the handling a bit more so the code is aware of different DBs , we will see how it looks. I think in the coq-lang 1.0 version we will be a bit limited on what we can do; hopefully once we revise the model and stop supporting coq_makefile we should be able to improve things.
